packages/protocol-identify/src/identify.ts

```diff
@@ -371,6 +252,180 @@ export class Identify implements Startable, IdentifyInterface {
       stream.abort(err)
     }
   }
+}
+
+export class IdentifyPush implements Startable, IdentifyPushInterface {
+  private readonly identifyPushProtocolStr: string
+  public readonly host: {
+    protocolVersion: string
+    agentVersion: string
+  }
+
+  private started: boolean
+  private readonly timeout: number
+  private readonly peerId: PeerId
+  private readonly peerStore: PeerStore
+  private readonly registrar: Registrar
+  private readonly connectionManager: ConnectionManager
+  private readonly addressManager: AddressManager
+  private readonly maxInboundStreams: number
+  private readonly maxOutboundStreams: number
+  private readonly maxIdentifyMessageSize: number
+  private readonly events: TypedEventTarget<Libp2pEvents>
+  private readonly runOnTransientConnection: boolean
+  private readonly log: Logger
+
+  constructor (components: IdentifyPushComponents, init: IdentifyPushInit = {}) {
+    this.started = false
+    this.peerId = components.peerId
+    this.peerStore = components.peerStore
+    this.registrar = components.registrar
+    this.addressManager = components.addressManager
+    this.connectionManager = components.connectionManager
+    this.events = components.events
+    this.log = components.logger.forComponent('libp2p:identify:push')
+
+    this.identifyPushProtocolStr = `/${init.protocolPrefix ?? defaultValues.protocolPrefix}/${MULTICODEC_IDENTIFY_PUSH_PROTOCOL_NAME}/${MULTICODEC_IDENTIFY_PUSH_PROTOCOL_VERSION}`
+    this.timeout = init.timeout ?? defaultValues.timeout
+    this.maxInboundStreams = init.maxInboundStreams ?? defaultValues.maxInboundStreams
+    this.maxOutboundStreams = init.maxOutboundStreams ?? defaultValues.maxOutboundStreams
+    this.maxIdentifyMessageSize = init.maxIdentifyMessageSize ?? defaultValues.maxIdentifyMessageSize
+    this.runOnTransientConnection = init.runOnTransientConnection ?? defaultValues.runOnTransientConnection
+
+    // Store self host metadata
+    this.host = {
+      protocolVersion: `${init.protocolPrefix ?? defaultValues.protocolPrefix}/${IDENTIFY_PROTOCOL_VERSION}`,
+      agentVersion: getAgentVersion(components.nodeInfo, init.agentVersion)
+    }
+
+    if ((init.runOnSelfUpdate ?? defaultValues.runOnSelfUpdate)) {
+      // When self peer record changes, trigger identify-push
+      components.events.addEventListener('self:peer:update', (evt) => {
+        void this.push().catch(err => { this.log.error(err) })
+      })
+    }
+  }
+
+  isStarted (): boolean {
+    return this.started
+  }
+
+  async start (): Promise<void> {
+    if (this.started) {
+      return
+    }
+
+    await this.peerStore.merge(this.peerId, {
+      metadata: {
+        AgentVersion: uint8ArrayFromString(this.host.agentVersion),
+        ProtocolVersion: uint8ArrayFromString(this.host.protocolVersion)
+      }
+    })
+
+    await this.registrar.handle(this.identifyPushProtocolStr, (data) => {
```

Comment:
```suggestion
    await this.registrar.handle(this.protocolStr, (data) => {
```
could help if we were to consolidate logic into a base class